apps/workshop/projects/SpoolV2/modules/spool.ts

```diff
@@ -0,0 +1,112 @@
+import type { ModuleDefinitionInterface } from '@defiyield/sandbox';
+import { GET_VAULTS_QUERY, GQL_HEADERS, Vault } from '../helpers/gql';
+import { FetchPoolsContext } from '../../../../sandbox/src/types/module';
+import SPOOL_ABI from '../abis/abi.json';
+import * as dotenv from 'dotenv';
+import { ethers } from 'ethers';
+
+export const SpoolModule: ModuleDefinitionInterface = {
+  name: 'SpoolModule',
+  chain: 'ethereum',
+  type: 'pools',
+
+  async preloadTokens(ctx) {
+    dotenv.config({ path: `${__dirname}\\.env` });
+    const allTokens = await ctx.axios({
+      url: process.env['DEFIYIELD_GRAPHQL_ENDPOINT'],
+      method: 'POST',
+      headers: GQL_HEADERS,
+      data: {
+        query: GET_VAULTS_QUERY,
+      },
+    });
+
+    const tokenIds: string[] = [];
+    allTokens.data.data.smartVaults.forEach((smartVault: Vault) => {
+      smartVault.assetGroup.assetGroupTokens.forEach((assetGroupToken) => {
+        const tokenId = assetGroupToken.token.id;
+        tokenIds.push(tokenId);
+      });
+    });
+    const uniqueTokenIds: string[] = [];
+    tokenIds.forEach((item) => {
+      if (uniqueTokenIds.indexOf(item) === -1) {
+        uniqueTokenIds.push(item.toLowerCase());
+      }
+    });
+
+    return uniqueTokenIds;
+  },
+
+  async fetchPools(ctx) {
+    const smartVaultsData = await ctx.axios({
+      url: process.env['DEFIYIELD_GRAPHQL_ENDPOINT'],
+      method: 'POST',
+      headers: GQL_HEADERS,
+      data: {
+        query: GET_VAULTS_QUERY,
+      },
+    });
+
+    const smartVaults = smartVaultsData.data.data.smartVaults;
+    const result = [];
+
+    for (let i = 0; i < smartVaults.length; i++) {
+      const item = smartVaults[i];
+      const poolValue = [];
+
+      const assetGroupTokens = item.assetGroup.assetGroupTokens;
+
+      const provider = await new ethers.providers.JsonRpcProvider(process.env['ETHEREUM_URL']);
+      const contract = await new ethers.Contract(
+        process.env['SPOOL_LENS_ADDRESS']!,
+        SPOOL_ABI,
+        provider,
+      );
+      const tokenAmounts = await contract.callStatic.getSmartVaultAssetBalances(item.id, false);
+
+      //TODO: replace above example with the one below
```

Comment:
Would be nice if we could take a look at that together on how to use their provider. Will DM you once i finish everything so we can discuss.